include/cuco/static_multimap_ref.cuh

```diff
@@ -0,0 +1,301 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/open_addressing/open_addressing_ref_impl.cuh>
+#include <cuco/hash_functions.cuh>
+#include <cuco/operator.hpp>
+#include <cuco/probing_scheme.cuh>
+#include <cuco/storage.cuh>
+#include <cuco/types.cuh>
+#include <cuco/utility/cuda_thread_scope.cuh>
+
+#include <cuda/std/atomic>
+
+namespace cuco {
+
+/**
+ * @brief Device non-owning "ref" type that can be used in device code to perform arbitrary
+ * operations defined in `include/cuco/operator.hpp`
+ *
+ * @note Concurrent modify and lookup will be supported if both kinds of operators are specified
+ * during the ref construction.
+ * @note cuCollections data structures always place the slot keys on the left-hand
+ * side when invoking the key comparison predicate.
+ * @note Ref types are trivially-copyable and are intended to be passed by value.
+ * @note `ProbingScheme::cg_size` indicates how many threads are used to handle one independent
+ * device operation. `cg_size == 1` uses the scalar (or non-CG) code paths.
+ *
+ * @throw If the size of the given key type is larger than 8 bytes
+ * @throw If the size of the given payload type is larger than 8 bytes
+ * @throw If the given key type doesn't have unique object representations, i.e.,
+ * `cuco::bitwise_comparable_v<Key> == false`
+ * @throw If the given payload type doesn't have unique object representations, i.e.,
+ * `cuco::bitwise_comparable_v<T> == false`
+ * @throw If the probing scheme type is not inherited from `cuco::detail::probing_scheme_base`
+ *
+ * @tparam Key Type used for keys. Requires `cuco::is_bitwise_comparable_v<Key>` returning true
+ * @tparam T Type used for mapped values. Requires `cuco::is_bitwise_comparable_v<T>` returning true
+ * @tparam Scope The scope in which operations will be performed by individual threads.
+ * @tparam KeyEqual Binary callable type used to compare two keys for equality
+ * @tparam ProbingScheme Probing scheme (see `include/cuco/probing_scheme.cuh` for options)
+ * @tparam StorageRef Storage ref type
+ * @tparam Operators Device operator options defined in `include/cuco/operator.hpp`
+ */
+template <typename Key,
+          typename T,
+          cuda::thread_scope Scope,
+          typename KeyEqual,
+          typename ProbingScheme,
+          typename StorageRef,
+          typename... Operators>
+class static_multimap_ref
+  : public detail::operator_impl<
+      Operators,
+      static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>>... {
+  /// Flag indicating whether duplicate keys are allowed or not
+  static constexpr auto allows_duplicates = true;
+
+  /// Implementation type
+  using impl_type = detail::
+    open_addressing_ref_impl<Key, Scope, KeyEqual, ProbingScheme, StorageRef, allows_duplicates>;
+
+  static_assert(sizeof(T) == 4 or sizeof(T) == 8,
+                "sizeof(mapped_type) must be either 4 bytes or 8 bytes.");
```

Comment:
I thought `cuda::atomic` would support 2 byte types by now

Reply:
https://godbolt.org/z/a78x4qrha wow, good to know. This unblocks many libcudf cleanups as well.

Reply:
I will create a separate PR to address this.

Reply:
https://github.com/NVIDIA/cuCollections/issues/554